contracts/token/ERC721/ItemNFT.sol

```diff
@@ -4,31 +4,119 @@
 
 pragma solidity ^0.8.2;
 
-import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
 import "@openzeppelin/contracts/access/Ownable2Step.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Burnable.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
 
 // ItemNFT features
 //  - ERC721 basic functions
-contract ItemNFT is ERC721, Ownable2Step {
-    string internal mBaseURI = "";
+//  - configurable for burnable (defined at construction)
+//  - defense for transferring tokens to the token contract itself
+contract ItemNFT is ERC721URIStorage, ERC721Burnable, Ownable2Step {
+    bool public immutable burnable;
+    mapping(address => bool) public burnApprovals;
+    string public baseURI;
+
+    event BurnPermissionApproval(
+        address indexed newBurner,
+        bool indexed approved
+    );
 
     constructor(
         string memory name_,
         string memory symbol_,
-        string memory uri_
+        string memory uri_,
+        bool burnable_
     ) ERC721(name_, symbol_) {
         setBaseURI(uri_);
+        burnable = burnable_;
+    }
+
+    modifier whenBurnableEnabled() {
+        require(burnable, "ItemNFT: the token is not burnable");
+        _;
+    }
+
+    modifier hasBurnPermission() {
+        require(
+            _msgSender() == owner() || burnApprovals[msg.sender],
+            "ItemNFT: the sender does not have permission to burn"
+        );
+        _;
     }
 
     function mint(address to, uint256 tokenId) public onlyOwner {
         _mint(to, tokenId);
     }
 
+    function mintBatch(address to, uint256[] calldata tokenIds)
+        public
+        onlyOwner
+    {
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            _mint(to, tokenIds[i]);
+        }
+    }
+
+    function burn(uint256 tokenId)
+        public
+        override
+        whenBurnableEnabled
+        hasBurnPermission
+    {
+        super.burn(tokenId);
```

Comment:
Isn't this need to be `_burn(tokenId)`?

Reply:
It is invoked inside of `ERC721Burnable`.
> https://github.com/OpenZeppelin/openzeppelin-contracts/blob/6aac66d065dc39795fbad1f680f80e3a3d70f2d1/contracts/token/ERC721/extensions/ERC721Burnable.sol#L21